RFC-82: Implement Support for Cedar Tags

adowair opened the pull request.

adowair committed:
types: add Tags field to Entity struct

Signed-off-by: Ali Dowair <ali.dowair@strongdm.com>

adowair committed:
internal/eval: evaluate tags

Signed-off-by: Ali Dowair <ali.dowair@strongdm.com>

adowair committed:
ast: implement tag nodes in AST

Signed-off-by: Ali Dowair <ali.dowair@strongdm.com>

adowair committed:
internal/parser: test parsing of tags in condition

Signed-off-by: Ali Dowair <ali.dowair@strongdm.com>

adowair committed:
types: fixup and test tag JSON marshalling

Signed-off-by: Ali Dowair <ali.dowair@strongdm.com>

philhassey commented on internal/eval/evalers.go:
```diff
@@ -804,6 +805,78 @@ func (n *hasEval) Eval(env Env) (types.Value, error) {
+	}
+
+	e, ok := env.Entities[eid]
+	if !ok {
```
Gotta update for latest Getter stuff 👍

adowair replied:
Rebased, thanks!

philhassey commented on internal/eval/partial_test.go:
```diff
@@ -1238,6 +1238,30 @@ func TestPartialBasic(t *testing.T) {
+			ast.Principal().HasTag(ast.String("key")),
+			testutil.OK,
+		},
+		{
```
It would be nice to have folding examples here ... maybe there needs to be some extra data dumped into the context or something so that's possible? e.g. `context.entity` and then have some entity defined in the entities.

philhassey reviewed (changes requested):
Looks good!

- corpus tests
- verify 100% coverage
- update to work against latest

Once those are done, you can close this PR and make a new PR against the actual cedar-go repo for final review, etc.

patjakdev commented on types/entity.go:
```diff
@@ -33,10 +34,12 @@ func (e Entity) MarshalJSON() ([]byte, error) {
 		UID        ImplicitlyMarshaledEntityUID   `json:"uid"`
 		Parents    []ImplicitlyMarshaledEntityUID `json:"parents"`
 		Attributes Record                         `json:"attrs"`
+		Tags       Record                         `json:"tags"`
```
The [Rust implementation](https://github.com/cedar-policy/cedar/blob/dd6b9f0e280e4ed5eb1bea33891e5f254f7185b5/cedar-policy-core/src/ast/entity.rs#L369) actually elides the field if it's empty, (presumably for backwards compatibility?). To match the behavior, we could use the `omitempty` option here.

adowair replied:
~That's great to know (and it fixes some incompatibilities even in our codebase 😄)~

Since `Tags` is a struct, it cannot have an empty value, so `omitempty` won't be enough here. I _could_ let `Tags` be `*Record`, which would solve this problem right away, but that possibly introduces a host of other problems of mutability.

Aside: a new tag `omitzero` will be introduced in Go `v1.24`, which was made to address this exact scenario (https://github.com/golang/go/issues/45669) ([commit](https://go-review.googlesource.com/c/go/+/615676)).

patjakdev replied:
Mmm, right. Thanks for the reminder.

philhassey commented:
Also, update the README.md removing the line about how entity tags are missing.

adowair committed:
always instantiate Env.Entities

This commit will make changes to product code (and not test code),
to always instantiate Env.Entities to types.EntityMap{}, if it
has not been instantiated. This is a quick fix to safeguard against
nil pointer access Env.Entities when Entities is not set.

Signed-off-by: Ali Dowair <ali.dowair@strongdm.com>

adowair committed:
update and add some corpus tests

This commit will update the corpus test bundle to latest, and copies
one of the added corpus tests to TestCorpusRelated, to catch tag
related corner cases easier.

Signed-off-by: Ali Dowair <ali.dowair@strongdm.com>

adowair committed:
internal/eval: test partial evaluation of tags

This commit will add a new test to ensure the partial evaluation
of tag methods on context functions correctly.

Signed-off-by: Ali Dowair <ali.dowair@strongdm.com>

adowair commented:
Superseded by https://github.com/cedar-policy/cedar-go/pull/77
